rynlwhh iP

rynlwhh opened the pull request:
# NaegaBot

> “Simplicity is the soul of efficiency.” — *Austin Freeman*

**NaegaBot** is here to help you manage tasks efficiently and effortlessly! It’s:

- Text-based  
- Easy to learn  
- ~~Overwhelming~~ **Super fast** to use ⚡

To get started:

1. [Download it here](https://github.com/rynlwhh/ip).
2. Double-click the app to launch it.
3. Start managing your tasks with simple commands.
4. Let NaegaBot do the rest! 😄

### Features:

- Managing tasks
- Managing deadlines
- Reminders (coming soon)

If you’re a Java developer, you can use it to practice your Java skills. Here’s the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Rename main class, added greeting and echo.

Commit:
Add support for list

Commit:
Added mark as done feature and task class

Commit:
Added Todo, Deadline, Event classes

Commit:
Automated Text UI Testing

Commit:
Handle errors by including exception class and cleaned up code

Commit:
Added Delete function

RezwanAhmed123 commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,15 @@
+public class Event extends Task {
+    protected String from;
+    protected String to;
```
I like the simplicity of the names of the time variables. However, it may be abit more clearer if the variables were more specific such as the following:

```suggestion
    protected String fromDateTime;
    protected String toDateTime;
```

RezwanAhmed123 commented on src/main/java/Naega.java:
```diff
@@ -0,0 +1,150 @@
+        scanner.close();
+    }
+
+    private static void handleList(ArrayList<Task> tasks) {
+        System.out.println("____________________________________________________________");
+        System.out.println(" Here are the tasks in your list:");
+        for (int i = 0; i < tasks.size(); i++) {
+            System.out.println(" " + (i + 1) + "." + tasks.get(i));
+        }
+        System.out.println("____________________________________________________________");
+    }
+
+    private static void handleMark(ArrayList<Task> tasks, String userInput) throws NaegaException {
+        int taskIndex = parseTaskIndex(userInput, "mark ");
+        tasks.get(taskIndex).markAsDone();
+        System.out.println("____________________________________________________________");
+        System.out.println(" Nice! I've marked this task as done:");
+        System.out.println("   " + tasks.get(taskIndex));
+        System.out.println("____________________________________________________________");
+    }
+
+    private static void handleUnmark(ArrayList<Task> tasks, String userInput) throws NaegaException {
+        int taskIndex = parseTaskIndex(userInput, "unmark ");
+        tasks.get(taskIndex).markAsNotDone();
+        System.out.println("____________________________________________________________");
+        System.out.println(" OK, I've marked this task as not done yet:");
+        System.out.println("   " + tasks.get(taskIndex));
+        System.out.println("____________________________________________________________");
```
I like how these methods were named. Good job!

Perhaps common code could be extracted out tho.

RezwanAhmed123 commented on src/main/java/Naega.java:
```diff
@@ -0,0 +1,150 @@
+        System.out.println("____________________________________________________________");
+    }
+
+    private static void handleTodo(ArrayList<Task> tasks, String userInput) throws NaegaException {
+        String description = parseDescription(userInput, "todo ");
+        Task newTask = new Todo(description);
+        tasks.add(newTask);
+        printTaskAdded(tasks, newTask);
+    }
+
+    private static void handleDeadline(ArrayList<Task> tasks, String userInput) throws NaegaException {
+        String[] parts = parseDescription(userInput, "deadline ").split(" /by ");
+        if (parts.length < 2) {
+            throw new NaegaException("The deadline must include a '/by' keyword.");
+        }
+        Task newTask = new Deadline(parts[0], parts[1]);
+        tasks.add(newTask);
+        printTaskAdded(tasks, newTask);
+    }
+
+    private static void handleEvent(ArrayList<Task> tasks, String userInput) throws NaegaException {
+        String[] parts = parseDescription(userInput, "event ").split(" /from ");
+        if (parts.length < 2) {
+            throw new NaegaException("The event must include a '/from' keyword.");
+        }
+        String description = parts[0];
+        String[] timeParts = parts[1].split(" /to ");
+        if (timeParts.length < 2) {
+            throw new NaegaException("The event must include a '/to' keyword.");
+        }
+        Task newTask = new Event(description, timeParts[0], timeParts[1]);
+        tasks.add(newTask);
+        printTaskAdded(tasks, newTask);
+    }
```
Similar to the previous comment, perhaps, the common code could be extracted into a separate method?

RezwanAhmed123 commented on src/main/java/Naega.java:
```diff
@@ -0,0 +1,150 @@
+        while (true) {
+            userInput = scanner.nextLine();
+
+            try {
+                if (userInput.equals("bye")) {
+                    System.out.println("____________________________________________________________");
+                    System.out.println(" Bye. Hope to see you again soon!");
+                    System.out.println("____________________________________________________________");
+                    break;
+                } else if (userInput.equals("list")) {
+                    handleList(tasks);
+                } else if (userInput.startsWith("mark ")) {
+                    handleMark(tasks, userInput);
+                } else if (userInput.startsWith("unmark ")) {
+                    handleUnmark(tasks, userInput);
+                } else if (userInput.startsWith("todo ")) {
+                    handleTodo(tasks, userInput);
+                } else if (userInput.startsWith("deadline ")) {
+                    handleDeadline(tasks, userInput);
+                } else if (userInput.startsWith("event ")) {
+                    handleEvent(tasks, userInput);
+                } else if (userInput.startsWith("delete ")) {
+                    handleDelete(tasks, userInput);
+                } else {
+                    throw new NaegaException("I don't know how to handle that command.");
+                }
+            } catch (NaegaException e) {
+                System.out.println("____________________________________________________________");
+                System.out.println(" Oops! " + e.getMessage());
+                System.out.println("____________________________________________________________");
+            } catch (NumberFormatException | IndexOutOfBoundsException e) {
+                System.out.println("____________________________________________________________");
+                System.out.println(" Oops! Invalid task number.");
+                System.out.println("____________________________________________________________");
+            }
```
I like how mostly clean this block of code is! Any reason for using if-else instead of switch-case here though?

RezwanAhmed123 reviewed (commented):
LGTM, very readable except some nitpicky comments. Some repetition in code could be removed by extracting common code in multiple methods. Overall well done!

Commit:
Implemented save and load functionality for tasks

Commit:
Store deadline dates as a java.time.LocalDate (or java.time.LocalDateTime) in your task objects.

Commit:
Made the code more OOP.

Commit:
Organized the classes into suitable java packages.

Commit:
Organized the classes into suitable java packages.

Commit:
Organized the classes into suitable java packages.

Commit:
Merge branch 'add-gradle-support'

Commit:
Modified the build.gradle to point to the right class.

Commit:
Add JUnit tests to test the behavior of the code

Commit:
Add JUnit tests to test the behavior of the code

Commit:
Packaged the App as a JAR file

Commit:
Implemented find function to give users a way to find a task by searching for a keyword in the task description

Commit:
Add header comments to all non-private classes/methods, and non-trivial private methods

Commit:
Tweak the code to comply with a given coding standard

Commit:
Merge branch 'branch-Level-9'

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/Naega/Naega.java

Commit:
Add a GUI to the chatbot

Commit:
Fix time format issue in task storage and parsing

The current time format for deadlines and events in the task storage file is inconsistent. Times like ‘0200’ are being saved as ‘200’, which leads to parsing errors when reloading tasks.

This needs to change because the inconsistent format causes the application to fail when attempting to parse times without leading zeros. Using an incorrect format affects the ability of the program to read and write date-time values correctly, which leads to crashes.

Change the DateTimeFormatter to use the format yyyy-MM-dd HHmm for both saving and parsing tasks. Ensure that all times are stored in a consistent format with leading zeros, preventing parsing errors when reading tasks back from the file.

This solution is chosen because it enforces a standardized format for date and time, which avoids ambiguity and ensures compatibility with the LocalDateTime parser in Java. By using a format that includes leading zeros, the program can consistently handle and display dates and times without issues.

Also, update the task storage file to reflect the new format to avoid further parsing problems.

Commit:
Utilize Ui.showLoadingError and Ui.showWelcome methods in MainWindow for better user feedback

The showLoadingError() and showWelcome() methods in the Ui class were previously unused, resulting in missed opportunities to provide user feedback when tasks failed to load or when the application started.

This needs to change because displaying a welcome message upon startup enhances the user experience by introducing the application. Additionally, informing users of errors, such as task loading failures, is critical to ensure transparency and guide users in case of issues.

The showWelcome() method is now used in the MainWindow class’s initialize() method to display a welcome message when the application starts. The showLoadingErrorInGui() method was added to MainWindow and is used to handle task-loading failures, ensuring the error is displayed in the graphical interface rather than just the console.

This approach ensures that the GUI provides clear feedback to users at critical points—during startup and when errors occur—improving both usability and error handling in the application.

Commit:
Added showError method back

Commit:
Use assert feature to document important assumptions that should hold at various points in the code

Commit:
Improve code quality by removing unused code and changing Javadoc comments to make it more accurate

Commit:
Improve code quality by removing unused code

Commit:
Improve code quality by removing unused code

rynlwhh committed:
Merge pull request #2 from rynlwhh/branch-A-Assertions

 Branch-A-Assertions

rynlwhh committed:
Merge pull request #3 from rynlwhh/branch-A-CodeQuality

branch-A-CodeQuality

Commit:
Changed the name of the Gui class to fit coding standards

Commit:
Provide in-App guidance to users

Commit:
Change GUI to include Naega

Commit:
Change GUI to include Naega

Commit:
Add a User Guide

Commit:
Fix loading sample data twice error

Commit:
Fix find command

Commit:
Fix build.gradle dependencies

Commit:
Change user and naega image in GUI

Commit:
Fix jar file so it launches properly

Commit:
Change ExitCommand so Bye input exits app

Commit:
Amended the user guide

Commit:
Amend the user guide

Commit:
Remove unecessary code

Commit:
Remove unecessary line

Commit:
Simplify code to make it more straightforward

Commit:
Remove unused import statement
